pkg/dnscache/dnscache.go

```diff
@@ -0,0 +1,204 @@
+package dnscache
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/v2/expirable"
+
+	"github.com/aquasecurity/tracee/pkg/utils/set"
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+var (
+	ErrDnsCacheIsNotInitialized = errors.New("dns cache is not initialized")
+)
+
+const (
+	dnsCacheSize        = 500
+	dnsCacheTTL         = time.Minute * 5
+	reverseSearchSuffix = ".in-addr.arpa"
+)
+
+type DNSCache struct {
+	ipToDNS *expirable.LRU[string, *set.Set[string]]
+	dnsToIP *expirable.LRU[string, *set.Set[string]]
+
+	rwLock *sync.RWMutex
+}
+
+func New() *DNSCache {
+	return &DNSCache{
+		ipToDNS: expirable.NewLRU[string, *set.Set[string]](dnsCacheSize, nil, dnsCacheTTL),
+		dnsToIP: expirable.NewLRU[string, *set.Set[string]](dnsCacheSize, nil, dnsCacheTTL),
+		rwLock:  &sync.RWMutex{},
+	}
+}
+
+// Add adds the DNS response into the DNS' LRU cache.
+func (dc *DNSCache) Add(dnsResponse trace.DnsResponseData) error {
+	switch dnsResponse.QueryData.QueryType {
+	case "A":
+	case "AAAA":
+	case "PTR":
+	default:
+		return nil // Do not handle other query types
+	}
```

Comment:
I wont ask you to rewrite using dnsRequest, but why didnt you use it ? I mean, the DnsRequest is the generic event that should always exist (instead of Request/Response events that were added only to keep compatibility with existing signatures back then).

Reply:
Because I didn't want to filter (again, as it's already filtered in the derivation) in the function wether it was a request or response.

Reply:
I believe this conversation thread was elucidated today in our talk. I wanted to make sure all types were covered when I asked this (and I thought that CNAME + A would come together in the response). This might or might not be true. And I think we already covered all the other cases today.

Reply:
> (and I thought that CNAME + A would come together in the response)

From what I've gathered, it seems to be true for "A" requests, servers send the CNAME and then the A records corresponding to it, but it probably depends on the server. I think the current implementation should usually cover us, but I may want to further change how DNS records are linked to each other before merging.

Reply:
Depends on the query as well. "set q=a" would send a response with all together.